### PR TITLE
MenuItemReview post (all users)

### LINKED ISSUE #7: Create Controller, add controller methods to post reviews (all users), get all reviews (admin only) (EPIC #2)
Subtask of #2 

# Dependencies

This issue should be completed after the issue with the title "Create database table for reviews." #6 

# Backend 2: Controller methods to post reviews (all users), get all reviews (admin only)

# Ex

https://github.com/ucsb-cs156-f24/team01-f24-15/issues/9

You'll probably need:
- controller file
- GET endpoint
- POST endpoint
- make sure that these endpoints have sufficient documentation so we know what it does
- GET and POST work on localhost
- GET and POST work on dokku
- full test coverage (Jacoco)
- full mutation test coverage (Pitest)

# What to do next

When you've met all of the acceptance criteria:

**Make a PR**: Do a PR for this branch, being sure to enter a [reasonable title](https://ucsb-cs156.github.io/topics/pull_requests.html#pr-titles)
and [description](https://ucsb-cs156.github.io/topics/pull_requests.html#pr-descriptions).

In the description, include the text "Closes #n" where `n` is the number of
this issue.

Then ask for a team member to review your PR.

Finally: drag this issue from "In Progress" to "In Review".

**Review Others' PRs**: Next, see if there are any PRs that need to be code reviewed and merged.  Unless you are the
first person on your team to complete a PR, there likely is at least one.
See: 
* [General Code Review Advice](https://ucsb-cs156.github.io/topics/code_reviews/)
* [Performin Code Reviews on Github](https://ucsb-cs156.github.io/topics/code_reviews/code_reviews_github.html)

**Start your next issue **: 

Now you can work on the next controllers issues.

Find the issue you are going to work on next on the Kanban board.

* Assign it to yourself
* Drag into "In Progress"
* Create a new branch, starting from the branch for the issue you just finished.
* Start on the new issue.

### PR DESCRIPTION
**Pull request by sophiaray2101**
Closes #7
dokku link: https://proj-dining-sophiaray2101.dokku-15.cs.ucsb.edu/

- [x] Created the controller and test files for post
- [x] 100% jacoco test coverage
- [x] 100% mutation coverage
- [x] works on both localhost and dokku

What to test:
1. Make sure there are no internal server errors trying to post
2. Make sure that if you are not logged in you cannot post
3. Check that there is a 404 error when trying to post to a non existent id.

When trying to do a post, I would first do the GET on the menu item to ensure that you know which item id you are leaving a review for and that the id's match

*NEW CHANGES*
1. there were too many issues with the get all and it wouldnt work no matter what I did, so I am just focusing on the post
2. CreatedDate and LastEditedDate are now automatically set as the local time at that moment. in the future when there are edits to be made, the LastEditedDate can be updated in that controller method.
3. StudentUserId is grabbed from the logged in user automatically.
5. Status is initally assigned as "Awaiting Moderation".
6. Moderator comments and userId are both not set.
7. Added ratings and reviewText to the entity and all other necessary files

Screenshot 2024-11-25 at 9 37 45 PM

### CODE REVIEW

**Comment by riasinha** [comment]
LGTM - tested get all on dokku for admin user and forbidden user and works as expected

**Review by riasinha — DISMISSED** [review]
LGTM - tested on dokku as admin and restricted user

File: src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    
+    /**
+     * Create a new menu item review -> all users
+     */
+
+    @Operation(summary= "Create a new menu item review")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping("/post")
+    public MenuItemReview postMenuItemReview(
+            @Parameter(name="studentUserId") @RequestParam long studentUserId,
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java]:
```suggestion

```

Instead, assign `studentUserId = currentUser.getId();` inside the controller

File: src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            @Parameter(name="itemId") @RequestParam long itemId,
+            @Parameter(name="itemServedDate", description="date (in iso format, e.g. YYYY-mm-ddTHH:MM:SS") @RequestParam("itemServedDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime itemServedDate,
+            @Parameter(name="status") @RequestParam String status,
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java]:
```suggestion
          
```
Instead, assign status = "Awaiting Moderation"

File: src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            @Parameter(name="moderatorUserId") @RequestParam long moderatorUserId,
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java]:
```suggestion

```

We don't know who the moderator will be.  That is determined when a moderator brings up the list of reviews awaiting moderation and clicks on one.  

Instead, assign 1moderatorId = null;` inside the controller

File: src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            @Parameter(name="moderatorComments") @RequestParam String moderatorComments,
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java]:
```suggestion

```

How could the user possibly know what the moderators comments are going to be?  Assign to null inside the controller.

File: src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            @Parameter(name="createdDate", description="date (in iso format, e.g. YYYY-mm-ddTHH:MM:SS") @RequestParam("createdDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime createdDate,
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java]: This should be set to the current date and time (`now`) inside the controller, or better yet, assigned using the techniques described here:

https://ucsb-cs156.github.io/topics/spring_boot/spring_boot_database.html#timestamps-on-database-rows

File: src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            @Parameter(name="lastEditedDate", description="date (in iso format, e.g. YYYY-mm-ddTHH:MM:SS") @RequestParam("lastEditedDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime lastEditedDate)
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java]: You forgot the two most important things!!!! 

The whole point of creating a review is to give the item a number of stars (1-5) and write some text about how much you loved or hated the food!

Where are those two parameters?

Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java]:
```suggestion
           
```

Last edited date should be set either automatically using the techniques here:

https://ucsb-cs156.github.io/topics/spring_boot/spring_boot_database.html#timestamps-on-database-rows

Or it should be set inside the controller in the POST and PUT methods, plus any method that the moderator uses to change the status.

Allowing the *user* to input this means they can choose any date/time they want, in the past, present or future.  What would be the purpose of the field if the user can make it say anything they want?

It's so important to keep the use case and the user story in mind when writing code for real applications!

File: src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     * List all menu item reviews -> ADMIN ONLY
+     */
+    @Operation(summary= "List all menu item reviews")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @GetMapping("/all")
+    public Iterable<MenuItemReview> allMenuItemReviewAdminOnly() {
+        Iterable<MenuItemReview> reviews = menuItemReviewRepository.findAll();
+        return reviews;
+    }
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/dining/controllers/MenuItemReviewController.java]: What's the use case for this?

I do understand why an admin would need to list all reviews that have the status "Awaiting Moderation". But when would they just need to see a list of all the reviews?

**Review by pconrad — CHANGES_REQUESTED** [review]
Some crucial changes that need to be made here.

**Review by pconrad — DISMISSED** [review]
LGTM... will test on dokku

**Review by pconrad — APPROVED** [review]
LGTM... will test on dokku

**Pull request merged**